Resilient super method calls

slavapestov opened the pull request:
Fixes <https://bugs.swift.org/browse/SR-3928>, <rdar://problem/31411193>.

slavapestov committed:
IRGen: Add mangling for method lookup functions

slavapestov committed:
Runtime: Introduce swift_lookUpClassMethod()

slavapestov committed:
IRGen: Emit method lookup function

slavapestov committed:
IRGen: Use method lookup function for resilient super method calls

Fixes <https://bugs.swift.org/browse/SR-3928>, <rdar://problem/31411193>.

slavapestov commented:
@jrose-apple If you're curious you can take a look at this now -- it works in manual testing and all execution tests pass, but I need to update IRGen tests and write an Evolution test.

jrose-apple commented on include/swift/Demangling/DemangleNodes.def:
```diff
@@ -119,6 +119,7 @@ NODE(MergedFunction)
 NODE(Metatype)
 NODE(MetatypeRepresentation)
 NODE(Metaclass)
+NODE(MethodLookupFunction)
```
Don't forget to update ABI.rst too!

slavapestov replied:
Old-timer alert: it's actually ABI/Mangling.rst these days

jrose-apple commented on include/swift/Runtime/RuntimeFunctions.def:
```diff
@@ -773,6 +773,17 @@ FUNCTION(InitClassMetadata,
               SizeTy->getPointerTo()),
          ATTRS(NoUnwind))
 
+// void swift_lookUpClassMethod(Metadata *metadata,
```
This mock declaration isn't right if it's supposed to return something!

jrose-apple commented on lib/IRGen/GenMeta.cpp:
```diff
@@ -1314,6 +1314,11 @@ namespace {
         return;
 
+      // Only emit a method lookup function if the class is resilient
+      // and has a non-empty vtable.
```
It only needs this if it's `open`, right?

slavapestov replied:
Unfortunately no because methods in extensions can also use super.

slavapestov replied:
(I will add tests for that too)

jrose-apple replied:
Er, right. And I thought explicitly about that on the other one.

jrose-apple commented on lib/IRGen/IRGenSIL.cpp:
```diff
@@ -5509,6 +5510,43 @@ void IRGenSILFunction::visitSuperMethodInst(swift::SuperMethodInst *i) {
+    // Load the superclass of the instance class resiliently.
+    //
+    // FIXME: Perhaps if we knew that the instance type of the base value
+    // was exact, we could avoid this.
```
Specifically, we can avoid this part when in the same resilience domain as the *current* class. But that's an optimization, so it can come later.

slavapestov replied:
Yeah, the instanceTy here is the “current” class. I’ll clarify the comments

slavapestov replied:
Ok, I get what you mean. The static type of 'self' might be in our resilience domain in which case we can assume what its superclass is, even if the ancestor that _defines_ the method is in a different resilience domain. This is simple enough to implement; I'll do it in this PR.

slavapestov committed:
IRGen: Don't visit class members when building resilient class metadata layouts

If the class is resilient to us, we cannot assume anything about the
layout of its member area, so let's not even try; just skip visiting
the members entirely.

slavapestov committed:
IRGen: Fix case where we would go through the vtable entry and not the dispatch thunk

Make sure we check resilience of the class defining the method,
and not the class with the override we're calling.

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test

slavapestov commented:
@swift-ci Please smoke test macOS

slavapestov commented:
@swift-ci Please smoke test macOS

slavapestov commented:
@swift-ci Please smoke test macOS

slavapestov commented:
@swift-ci Please smoke test macOS

slavapestov commented:
@swift-ci Please smoke test macOS

slavapestov commented:
@swift-ci Please test macOS

slavapestov commented:
@swift-ci Please test macOS
